source/agora/script/Signature.d

```diff
@@ -142,39 +213,147 @@ unittest
         tx = the transaction to sign
         sig_hash = the `SigHash` to use
         input_idx = the associated input index we're signing for
+        output_idx = the associated output index we're signing for
 
     Returns:
         the challenge as a hash
 
 *******************************************************************************/
 
 public Hash getChallenge (in Transaction tx, in SigHash sig_hash = SigHash.All,
-    in ulong input_idx = 0) nothrow @safe
+    in ulong input_idx = 0, in ulong output_idx = 0) nothrow @safe
 {
+    import agora.script.Lock;
+    import std.exception : assumeWontThrow;
+
+    if (sig_hash != SigHash.All)
+    {
+        assert(input_idx < tx.inputs.length, "Input index is out of range");
+        if (sig_hash & SigHash.Single)
+            assert(output_idx < tx.outputs.length, "Output index is out of range");
+    }
+
+    Transaction dup = () @trusted { return *cast(Transaction*)&tx; }();
+
     final switch (sig_hash)
     {
     case SigHash.NoInput:  // eltoo support
-        assert(input_idx < tx.inputs.length, "Input index is out of range");
-        Transaction dup;
-        // it's ok, we'll dupe the array before modification
-        () @trusted { dup = *cast(Transaction*)&tx; }();
-        dup.inputs = dup.inputs.dup;
-        dup.inputs[input_idx] = Input.init;  // blank out matching input
+        dup.inputs = dup.inputs[0..input_idx] ~ dup.inputs[input_idx + 1..$]; // blank out matching input
+        return hashMulti(dup, sig_hash);
+    // sign all inputs and a single output
+    case SigHash.Single:
+        dup.outputs = dup.outputs[output_idx .. output_idx + 1];
+        return hashMulti(dup, sig_hash);
+    // sign a single input and a single output
+    case SigHash.Single | SigHash.AnyoneCanPay:
```

Comment:
One reason why using OOB values for `enum` is bad: The compiler could decide that this is definitely unreachable. Not to mention, `final switch` is useless here.